texts/0000-applicative-monad-superclass.rst

```diff
@@ -0,0 +1,219 @@
+``MonadPlus``, as already implemented in GHC as of 7.10 (released in early
+2015).
+
+
+
+##########
+Motivation
+##########
+
+
+It's the right thing to do™
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Math. You've all heard this one, it's good and compelling so I don't need to
+spell it out.
+
+
+
+Redundant functions
+~~~~~~~~~~~~~~~~~~~
+
+- ``pure`` and ``return`` do the same thing.
+- ``>>`` and ``*>`` are identical.
+- ``liftM`` and ``liftA`` are ``fmap``. The ``liftM*`` are ``liftA*``, ``<*>``
+  is ``ap``.
+- Prelude's ``sequence`` requres ``Monad`` right now, while ``Applicative`` is
+  sufficient to implement it. The more general version of this issue is
+  captured by ``Data.Traversable``, whose main typeclass implements the *same*
+  functionality twice, namely ``traverse`` and ``mapM``, and ``sequenceA`` and
+  ``sequence``.
+
+That very much violates the “don't repeat yourself” principle, and even more so
+it *forces* the programmer to repetition in order to achieve maximal generality.
+
+
+
+
+Using Functor/Applicative functions in monadic code
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Whenever there's Monad code, you can use ``Functor``/``Applicative`` functions,
+without introducing an additional constraint. Keep in mind that
+“``Functor``/``Applicative`` functions” does not only include what their
+typeclasses define but many more, for example ``void``, ``(<$>)``, ``(<**>)``.
+
+Even if you think you have monadic code, strictly using the least restrictive
+functions may result in something that requires only ``Applicative``. This is
+similar to writing a function that needs ``Int``, but it turns out any
+``Integral`` will do – more polymorphism for free.
+
+
+
+Beginner friendliness
+~~~~~~~~~~~~~~~~~~~~~
+
+How often did you say …
+
+- “A Monad is always an Applicative but due to historical reasons it’s not but
+  you can easily verify it by setting ``pure = return`` and ``(<*>) = ap``”
+- “``liftM`` is ``fmap`` but not really.” – “So when should I use ``fmap`` and
+  when ``liftM``?” - *sigh*
+
+With the new hierarchy, the answer would *always* be “use the least restrictive
+one”.
+
+
+
+
+###############
+Detailed design
+###############
+
+Add the following definitions to the Prelude, and demand their existence in the
+Report:
+
+.. code-block:: haskell
+
+   -- (Unchanged)
+   class  Functor f  where
+
+       -- Minimal complete definition: fmap
+
+       fmap :: (a -> b) -> f a -> f b
+
+       (<$) :: a -> f b -> f a
+       (<$) =  fmap . const
+
+
+
+   -- (Unchanged)
+   class Functor f => Applicative f where
+
+       -- Minimal complete definition: pure and (<*>)
+
+       pure :: a -> f a
+
+       (<*>) :: f (a -> b) -> f a -> f b
+
+       (*>) :: f a -> f b -> f b
+       (*>) x y = const id <$> x <*> y
+
+       (<*) :: f a -> f b -> f a
+       (<*) x y = const <$> x <*> y
+
+
+
+   class Applicative m => Monad m where
+
+       -- Minimal complete definition: (>>=) or join
```

Comment:
I think we should mandate join being part of the Monad class.  The reason
it was left out is very ghc specific, and we should not let that stop us
from doing the right thing.


On Fri, Aug 19, 2016 at 5:42 AM, Carter Tazio Schonwald <
notifications@github.com> wrote:

> In texts/0000-applicative-monad-superclass.rst
> <https://github.com/haskell/rfcs/pull/1#discussion_r75424014>:
>
> > +
> > +       pure :: a -> f a
> > +
> > +       (<*>) :: f (a -> b) -> f a -> f b
> > +
> > +       (*>) :: f a -> f b -> f b
> > +       (*>) x y = const id <$> x <*> y
> > +
> > +       (<*) :: f a -> f b -> f a
> > +       (<*) x y = const <$> x <*> y
> > +
> > +
> > +
> > +   class Applicative m => Monad m where
> > +
> > +       -- Minimal complete definition: (>>=) or join
>
> I believe join used to be a method of monad, but some limits to the work
> on recovering soundness in the process of generalized new type deriving
> necessitated moving it out. But having join be one path to defining monad
> again might be worth revisiting?
> … <#m_-5280049804865897887_>
> On Thursday, August 18, 2016, M Farkas-Dyck ***@***.***> wrote: In
> texts/0000-applicative-monad-superclass.rst <#1 (comment)
> <https://github.com/haskell/rfcs/pull/1#discussion_r75422700>>: > + > +
> pure :: a -> f a > + > + (<*>) :: f (a -> b) -> f a -> f b > + > + (*>) ::
> f a -> f b -> f b > + (*>) x y = const id <$> x <*> y > + > + (<*) :: f a
> -> f b -> f a > + (<*) x y = const <$> x <*> y > + > + > + > + class
> Applicative m => Monad m where > + > + -- Minimal complete definition:
> (>>=) or join join is no class method here. (I think it ought to be, but i
> also think that may be a topic for another RFC.) — You are receiving this
> because you are subscribed to this thread. Reply to this email directly,
> view it on GitHub <https://github.com/haskell/rfcs/pull/1/files/
> 6f0a803555b04cb205d764362c411867918f8ae5#r75422700>, or mute the thread <
> https://github.com/notifications/unsubscribe-auth/AAAQwqD-OC-
> PfGUKjXTiD5KmNlD2MLx1ks5qhSA5gaJpZM4JS1Lq> .
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/rfcs/pull/1/files/6f0a803555b04cb205d764362c411867918f8ae5#r75424014>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFAInkK13szoDLVvLScxiB87jaHzbmB1ks5qhSYSgaJpZM4JS1Lq>
> .
>